### PR TITLE
AddressBase Core import & searching functionality

### PR DESCRIPTION
**Pull request by davea**
- UPRN model to store rows from AddressBase Core
- Management command to import AddressBase Core CSV
- `/uprn` URL for viewing a single UPRN record as JSON or HTML (https://github.com/mysociety/societyworks/issues/2701)
- `/addressbase` URL for querying AddressBase Core (https://github.com/mysociety/societyworks/issues/2720)

(More tests for this functionality are included in a [subsequent PR](https://github.com/mysociety/labour-mapit/pull/3))

### CODE REVIEW

File: mapit_labour/management/commands/mapit_labour_import_addressbase_core.py [hunk 1]
```diff
@@ -104,4 +104,7 @@ def create_uprn(self, row: Dict[str, str]):
 
     def print_stats(self):
         c = self.count
-        print(f"Imported {c['total']} ({c['created']} new, {c['updated']} updated)")
+        print(
+            f"Imported {c['total']} ({c['created']} new, {c['updated']} updated)",
+            file=self.stdout,
+        )
```
Comment (dracos) [inline mapit_labour/management/commands/mapit_labour_import_addressbase_core.py:159]: https://docs.djangoproject.com/en/3.2/howto/custom-management-commands/#management-commands-output recommends self.stdout.write (I assume it's the same, though).

File: mapit_labour/admin.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            try:
+                cursor = connection.cursor()
+                cursor.execute(
+                    "SELECT reltuples FROM pg_class WHERE relname = %s",
+                    [query.model._meta.db_table],
+                )
+                return int(cursor.fetchone()[0])
+            except:  # pragma: no cover
+                pass
+        return super(LargeTablePaginator, self).count
```
Comment (dracos) [inline mapit_labour/admin.py]: super() as we're python3? Doesn't matter, though.

File: mapit_labour/management/commands/mapit_labour_import_addressbase_core.py [hunk 3]
```diff
@@ -44,40 +44,30 @@ def handle(self, *args, **kwargs):
...
-                uprn=uprn,
-                defaults=dict(
-                    postcode=postcode,
-                    location=location,
-                    addressbase=row,
-                ),
-            )
-            self.count["created" if created else "updated"] += 1
         self.count["total"] += 1
+        self.count["created"] += 1
```
Comment (dracos) [inline mapit_labour/management/commands/mapit_labour_import_addressbase_core.py:103]: Does it need to have both total and created? The stats output doesn't seem accurate now.

File: mapit_labour/management/commands/mapit_labour_import_addressbase_core.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        super().handle(*args, **kwargs)
+
+    def handle_rows(self, csv: DictReader):
+        if self.purge:
+            UPRN.objects.all().delete()
+
+            for rows in batched(islice(csv, self.limit or None), self.batch_size):
+                with transaction.atomic():
+                    UPRN.objects.bulk_create(self.create_uprn(row) for row in rows)
```
Comment (dracos) [inline mapit_labour/management/commands/mapit_labour_import_addressbase_core.py]: If purge isn't supplied, this management command doesn't do anything, I think?

**Review by dracos — CHANGES_REQUESTED** [review]
Couple of tiny things, plus I don't think the management command works without purge (after the final changes). I'd squash all these together presumably, no need for migrations to appear than disappear or the changes in the model, and the in-between bits of the management command creation.

File: mapit_labour/management/commands/mapit_labour_import_addressbase_core.py (hunk 4 above)
Comment (davea) [reply mapit_labour/management/commands/mapit_labour_import_addressbase_core.py]: You're right! I've implemented that in 3e3a8ef6521828ad05cfb06dbba15510c3865b63 in a (hopefully) performant way.

File: mapit_labour/management/commands/mapit_labour_import_addressbase_core.py (hunk 3 above)
Comment (davea) [reply mapit_labour/management/commands/mapit_labour_import_addressbase_core.py:103]: Now we can update existing records this is more useful. Total is number of rows processed from the CSV, and is a sum of created and updated.

**Comment by davea** [comment]
Thanks for the nudge to implement the non-`--purge` behaviour!

File: mapit_labour/management/commands/mapit_labour_import_addressbase_core.py [hunk 5]
```diff
@@ -102,6 +109,54 @@ def create_uprn(self, row: Dict[str, str]):
...
+    def find_existing_uprns(self, rows):
+        """
+        Takes a list of rows from the CSV and divides them into two lists which
+        are returned: those that don't already exist as URPN objects in the DB
+        and those that do.
+        """
+        # need to consume this iterator now because we use it twice below
+        rows = list(rows)
+        db_uprns = set(
+            UPRN.objects.filter(uprn__in=(r["UPRN"] for r in rows)).values_list(
```
Comment (dracos) [inline mapit_labour/management/commands/mapit_labour_import_addressbase_core.py]: The result of the filter here, is the same as the filter in `update_existing_uprns` which you're going to do anyway, so can we do it here and store it for there? It *is* `existing` basically, isn't it, once you have it, you only need to then construct `new` from the ones that aren't in it. Or have I misread.

**Review by dracos — COMMENTED** [review]
Looks good; I think there could still be a bit to improve the UPRN import, but it's late and I m might have got it wrong.

**Comment by dracos** [comment]
As I was in this codebase, I've implemented my suggestion above as 32d6511 - @davea Can you take a look and check it looks okay? Not sure if this is better/worse from memory point of view, guess it uses more for a new import but less if the data has already been imported, maybe?

File: mapit_labour/management/commands/mapit_labour_import_addressbase_core.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+    def find_existing_uprns(self, rows):
+        """
+        Takes a list of rows from the CSV and divides them into two lists which
+        are returned: those that don't already exist as URPN objects in the DB
+        and those that do.
+        """
+        # need to consume this iterator now because we use it twice below
+        rows = list(rows)
+        db_uprns = {u.uprn: u for u in UPRN.objects.filter(uprn__in=(r["UPRN"] for r in rows))}
```
Comment (davea) [inline mapit_labour/management/commands/mapit_labour_import_addressbase_core.py]: This change looks good to me, makes it a bit simpler doesn't it. I used `black` on this project so please format this line as so:

```suggestion
        db_uprns = {
            u.uprn: u for u in UPRN.objects.filter(uprn__in=(r["UPRN"] for r in rows))
        }
```

**Pull request merged**